src/apiserver/api_server_impl.cc

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright 2021 4Paradigm
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "api_server_impl.h"
+
+#include <memory>
+
+#include "brpc/server.h"
+#include "interface_provider.h"
+#include "json_writer.h"
+
+namespace fedb {
+namespace http {
+
+APIServiceImpl::~APIServiceImpl() = default;
+
+bool APIServiceImpl::Init(const sdk::ClusterOptions& options) {
+    // If cluster sdk is needed, use ptr, don't own it. SQLClusterRouter owns it.
+    auto cluster_sdk = new ::fedb::sdk::ClusterSDK(options);
+    bool ok = cluster_sdk->Init();
+    if (!ok) {
+        LOG(ERROR) << "Fail to connect to db";
+        return false;
+    }
+
+    auto router = std::make_unique<::fedb::sdk::SQLClusterRouter>(cluster_sdk);
+    if (!router->Init()) {
+        LOG(ERROR) << "Fail to connect to db";
+        return false;
+    }
+    sql_router_ = std::move(router);
+
+    RegisterPut();
+    RegisterExecSP();
+    RegisterGetSP();
+
+    return true;
+}
+
+void APIServiceImpl::Process(google::protobuf::RpcController* cntl_base, const HttpRequest*, HttpResponse*,
+                             google::protobuf::Closure* done) {
+    brpc::ClosureGuard done_guard(done);
+    auto* cntl = dynamic_cast<brpc::Controller*>(cntl_base);
+
+    // The unresolved path has no slashes at the beginning, it's not good for url parsing
+    auto unresolved_path = "/" + cntl->http_request().unresolved_path();
```

Comment:
check if the path starts with `/`

Reply:
brpc ensures that, cntl.http_request().unresolved_path() is always normalized, namely no slashes at the beginning or the end, and no repeated slashes in the middle. Ref https://github.com/apache/incubator-brpc/blob/master/docs/en/http_service.md#restful-url